### PR TITLE
Public Review: Procedures around Kotlin Language Evolution

### PR DESCRIPTION
**Pull request by abreslav**
This PR contains the initial version of the documents describing how Kotlin's evolution is governed. These documents describe the general guidelines on how decisions about language and core library changes are made.

To understand the context, please read the [langauge-committee.md](https://github.com/JetBrains/kotlin-web-site/blob/5688f9cd1ee14283423417b0b6f3a7bb1b5d5582/pages/docs/evolution/language-committee.md) file first.

These documents have been prepared by the Kotlin Language Committee, and are now offered here to gather feedback from the community. Your comments are very welcome

### CODE REVIEW

**Comment by jvgogh** [comment]
Looks good to me!

File: pages/docs/evolution/breaking-changes.md [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+* `4.2.1` Provide and describe a replacement for
+    deprecated constructs as part of the warning.
+
+* `4.2.2` Allow to disable the warnings in settings,
+    or promote them to errors.
+
+* `4.2.3` Along with this version, provide automated
+    migration aids (e.g. in the IDE).
+
+`4.3` Allow a long enough period of time for
```
Comment (voddan) [inline pages/docs/evolution/breaking-changes.md]: IMHO it is important to specify in this document the exact minimum migration period.

Many teams wait a certain time before updating to a new version of a software to allow for the vendor to release "hot bug fix" updates. I assume those teams use the same approach to the Kotlin compiler, IDE and IDE plugin. 

The problem arises when a team's migration period is significantly bigger than Kotlin's migrating period. In that case it may happen that a team skips several Kotlin compiler version, thus not following the intended migration path.

By specifying the minimum migration period for Kotlin components, for example 45 days, the Kotlin team guaranties that anybody can update their toolchain once per 45 days and not miss a version.

To clarify, this would still allow the length of the migration period to vary "based on the impact and complexity of the change".

File: pages/docs/evolution/breaking-changes.md [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+    discontinued.
+
+* `4.5.2` If possible, keep a compatibility mode that
+    does not support new features, but allows the deprecated one as in
+    version N.M+1 or N+1.X.
+
+* `4.5.3` Migration aids can be removed from the
+    compiler at this point.
+
+`4.6` A version that supports automated migration
```
Comment (voddan) [inline pages/docs/evolution/breaking-changes.md]: It is unclear which tools are covered by this paragraph. Since it mentions "automated migration", I assume that it includes not only the compiler, but also the IDE and the IDE plugin. In this case, it is unclear to me how it will work. Will JetBrains support installing older versions of the Kotlin plugin? Will the release notes contain "upgrade with x - y plugin version" notices? 

As a solution I can propose extracting the language migration functionality into separate plugins. Here is an example of how it would work: suppose I want to migrate my code base from Kotlin 1.6 to Kotlin 1.8; then I install the latest version of the Kotlin plugin, plus two smaller plugins: "Kotlin Migrator 1.6 -> 1.7" and "Kotlin Migrator 1.7 -> 1.8", where the additional plugins provide inspections and quick fixes needed to migrate from 1.6 to 1.7, and from 1.7 to 1.8 respectively. This configuration allows for different update cycles for the main plugin and for the migration assisting plugins. The downside is that it would probably require a somewhat stable internal compiler API to allow the plugins to communicate.

File: pages/docs/evolution/breaking-changes.md (hunk 1 above)
Comment (abreslav) [reply pages/docs/evolution/breaking-changes.md]: Language versions don't come out every 45 days, it takes closer to a year to release a new version. These documents are intended to set gudielines, not provide guaratees, so there's no need to write out exact numbers.

File: pages/docs/evolution/breaking-changes.md (hunk 2 above)
Comment (abreslav) [reply pages/docs/evolution/breaking-changes.md]: That's a good option

File: pages/docs/evolution/breaking-changes.md (hunk 1 above)
Comment (voddan) [reply pages/docs/evolution/breaking-changes.md]: My comment related to minor language versions rather than to major versions. The minor versions come more often than once a year, but I agree with you that their cooking period (~ 6 months) is probably more than enough.

File: pages/docs/evolution/breaking-changes.md [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+
+* `4.4.1` In the release notes, repeat the information
+    about the deprecation plan.
+
+* `4.4.2` Allow to demote the errors to warnings through
+    settings.
+
+* `4.4.3` Keep migration aids available.
+
+`4.5` In the version N.M+2 or N+1, report errors for
```
Comment (udalov) [inline pages/docs/evolution/breaking-changes.md]: ... or N+2?

File: pages/docs/evolution/breaking-changes.md (hunk 1 above)
Comment (pdvrieze) [reply pages/docs/evolution/breaking-changes.md]: From my perspective (university, teaching), it is important that I can continue using the same language version throughout the entire academic year while still having bug fixes for "significant" bugs (students are steered well away from edge cases and optimization is not an issue either). Changes in language versions throughout a course just increase opportunities for confusion as well as create problems with lab setup (which is often done once per year)

Comment (abreslav) [reply pages/docs/evolution/breaking-changes.md]: Release versions are kept updated with important big fixes, and you don't have to migrate to a next release even if it's come out already, so I don't think there should be any problems in your case

**Comment by artem-zinnatullin** [comment]
Can you please explicitly cover how new major version of Kotlin will work with previous one used in the same classpath/binary (and as a build system dependency)?

Will users have to update everything to new major version (project itself and all dependencies that use Kotlin) or new version will have own package ie `kotlin2` and users will be able to use both at runtime? See http://jakewharton.com/java-interoperability-policy-for-major-version-updates/

As the community grows, we have more and more libraries in Kotlin and migration from one major version to another could be very painful (as it is in Swift for instance).

Thanks!

**Comment by abreslav** [comment]
I think we are not ready to set a policy for such things yet, but we realize that binary compatibility issues are important and have to be accounted for. Why not now: nobody knows what happens with Jigsaw adoption and development, maybe the world of the classpath goes away by itself :)

File: pages/docs/evolution/change-review-process.md [hunk 4]
```diff
@@ -47,3 +37,42 @@ Reviews that need a decision before the next committee meeting (e.g.
 because of a release deadline) can be done by the committee in
 email/chat. These type of reviews need to be requested by the Lead
 Kotlin Language Designer (currently Andrey Breslav).
+
+# Issue Template
+
+``` java
```
Comment (LouisCAD) [inline pages/docs/evolution/change-review-process.md]: Why ```java?

Comment (abreslav) [reply pages/docs/evolution/change-review-process.md]: Because the website doesn't support markdown as a language yet

File: pages/docs/evolution/breaking-changes.md (hunk 3 above)
Comment (ilya-g) [reply pages/docs/evolution/breaking-changes.md]: or N+1.x?

File: pages/docs/evolution/change-review-process.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+# Proposed behavior/pattern
+> 1. How it fixes the issue?
+> 2. Describe automated migration, if possible
+
+# Impact of this change
+> If possible: 
+> * number of impacted users
+> * how common this pattern is
+> * if we believe it's a rare case, why
+> * is automated migration possible
```
Comment (ilya-g) [inline pages/docs/evolution/change-review-process.md]: "is automated migration possible" — is it the same as "Describe automated migration, if possible" in Proposed behavior?

**Comment by dnpetrov** [comment]
I'd suggest rewording
> 3.2.3 Bad code compiled due to a bug, and worked reasonably, must follow the deprecation policy
as
> 3.2.3 Bad code compiled and worked reasonably, must follow the deprecation policy

The actual reason why the code is considered "bad" might be more complex than just a "bug in the compiler". The point here is that some code that worked _reasonably_ (due to some coincidence) would behave differently after the change.

**Comment by dnpetrov** [comment]
Case study: https://youtrack.jetbrains.com/issue/KT-20830
Technically, there's no "bad" Kotlin code: here a Java method annotated as `NotNull` returns null.
However, in the end the result of this method is not used. No exception is thrown.

1.) Is this code "good" or "bad"?

2.) Does this code "work reasonably"?

We've discussed it face-to-face with @erokhins , and are kinda stuck.

**Comment by erokhins** [comment]
It is good to mention in `change-review-process.md` that for bugs that was introduced in the latest version language committee discussion can be established via e-mail.
It was shortly discussed at Mar 23rd, 2018: [doc](https://docs.google.com/document/d/1ReH84Cw_ZhGOUM_MdMQbLjzB0edXIeaFuBUF5molsuI/edit#heading=h.xmrj33x4ns7i)

**Comment by abreslav** [comment]
This was merged back in 2018, just not closed properly. Thanks everyone for your comments!
